### PR TITLE
Introduce and use a zipWithCollections() utility function

### PR DESCRIPTION
**Pull request by sschuberth**
Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: model/src/main/kotlin/OrtResult.kt [hunk 1]
```diff
@@ -200,10 +200,8 @@ data class OrtResult(
         val scannerIssues = scanner?.results?.collectIssues().orEmpty()
         val advisorIssues = advisor?.results?.collectIssues().orEmpty()
 
-        val analyzerAndScannerIssues =
-            analyzerIssues.zipWithDefault(scannerIssues, emptySet()) { left, right -> left + right }
-
-        return analyzerAndScannerIssues.zipWithDefault(advisorIssues, emptySet()) { left, right -> left + right }
+        val analyzerAndScannerIssues = analyzerIssues.zipWithCollections(scannerIssues)
+        return analyzerAndScannerIssues.zipWithCollections(advisorIssues)
```
Comment (fviernau) [inline model/src/main/kotlin/OrtResult.kt:204]: How about ditching the newline in 202 instead of the one above the return?

**Review by fviernau — CHANGES_REQUESTED** [review]

File: utils/src/main/kotlin/Extensions.kt [hunk 2]
```diff
@@ -273,6 +273,21 @@ inline fun <K, V, W> Map<K, V>.zipWithDefault(other: Map<K, V>, default: V, oper
...
+ * Merge two maps which have collections as values by creating the combined key set of both maps and merging the
+ * collections. If there is no entry for a key in one of the maps, the value from the other map is used.
+ */
+@Suppress("UNCHECKED_CAST")
+fun <K, V : Collection<T>, T> Map<K, V>.zipWithCollections(other: Map<K, V>): Map<K, V> =
+    zip(other) { a, b ->
+        when {
+            // When iterating over the combined key set, not both values can be null.
+            a == null -> b!!
+            b == null -> a
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/Extensions.kt:286]: Nope, Kotlin is smart enough here :-)

File: model/src/main/kotlin/OrtResult.kt (hunk 1 above)
Comment (sschuberth) [reply model/src/main/kotlin/OrtResult.kt:204]: I'd rather keep the two consecutive `zipWithCollections()` grouped.

File: utils/src/main/kotlin/Extensions.kt [hunk 3]
```diff
@@ -266,7 +266,7 @@ inline fun <K, V, W> Map<K, V>.zip(other: Map<K, V>, operation: (V?, V?) -> W):
 
 /**
  * Merge two maps by iterating over the combined key set of both maps and applying [operation] to the entries for the
- * same key. If there is no entry for a key in one of the maps, [default] is used.
+ * same key. If there is no entry for a key in one of the maps, [default] is used as the value for that map.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/Extensions.kt:269]: You mean in the first part of the sentence? But what does "respective" refer to then?

In the above function's docs "respective" made sense as it refers to "null".

File: utils/src/test/kotlin/ExtensionsTest.kt [hunk 4]
```diff
@@ -354,6 +354,29 @@ class ExtensionsTest : WordSpec({
...
+                "1" to listOf(1),
+                "2" to listOf(2),
+                "3" to listOf(3)
+            )
+            val other = mapOf(
+                "3" to listOf(3),
+                "4" to listOf(4)
+            )
+
+            val expectedResult = mapOf(
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/test/kotlin/ExtensionsTest.kt]: As I was copying this from the test above, I added a new commit to inline this for all.

File: utils/src/main/kotlin/Extensions.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply utils/src/main/kotlin/Extensions.kt:286]: ...in contrast to me ;-)

**Review by fviernau — APPROVED** [review]

**Pull request merged**